plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java

```diff
@@ -642,21 +648,324 @@ Invoker createInvoker() {
+                            }
+
+                            // Only report it in potential bread-crumb log if we
+                            // did not have a reference to this script/step/class
+                            // from the start of x.getMessage() effectively.
+                            // Note this is not a log line where we have a source
+                            // line number.
+                            if (!(xMsgStart.toString().contains(overflowedClassNameShort)))
+                                overflowedClassNameBreadcrumbs.append(l).append("\n");
+
+                            // Update the matching pattern in case we manage
+                            // to spot our problematic source in the stack trace
+                            CLASSNAME_MENTIONS_PATTERN = Pattern.compile("^\\s+at .*(WorkflowScript.*|Script\\d+|" + overflowedClassNameShort + ".*|\\.groovy):\\d+\\).*$");
+                            continue;
+                        } catch (Throwable ignored) {
+                        }
+                    }
+                }
+
+                Matcher matcher = CLASSNAME_MENTIONS_PATTERN.matcher(l);
+                if (matcher.find()) {
+                    overflowedClassNameBreadcrumbs.append(l).append("\n");
+                }
+            }
+        }
+
+        String overflowedClassNameReport;
+        if (overflowedClassName == null) {
+            overflowedClassNameReport = "WorkflowScript (the pipeline script) or one of its constituents";
+        } else if (overflowedClassName.equals("WorkflowScript") || CLASSNAME_SCRIPTNUM_PATTERN.matcher(overflowedClassName).find()) {
+            overflowedClassNameReport = overflowedClassName + " (the pipeline script)";
+        } else if (overflowedClassName.contains("/")) {
+            // quote the step/class name pretty:
+            //   FAILED to parse 'stepName' due to...
+            overflowedClassNameReport = "presumed JSL class '" + overflowedClassName + "'";
+        } else {
+            // quote the step/class name pretty:
+            //   FAILED to parse 'stepName' due to...
+            overflowedClassNameReport = "presumed JSL step '" + overflowedClassName + "'";
+        }
+
+        // Short message suffices, not much that a pipeline developer
+        // can do with the stack trace into the guts of groovy
+        StringBuilder actionableMsg = new StringBuilder();
+        actionableMsg
+                .append("FAILED to parse ")
+                .append(overflowedClassNameReport)
+                .append(" due to MethodTooLargeException");
+        if (ecCount > 1) {
+            actionableMsg.append(" (and other issues)");
+        }
+        actionableMsg.append("; please refactor to simplify code structure");
+        if (overflowedClassNameReport.contains("WorkflowScript") || CLASSNAME_SCRIPTNUM_PATTERN.matcher(overflowedClassName).find())
+            actionableMsg.append(" and/or move logic to a Jenkins Shared Library");
```

Comment:
BTW the terminology in https://www.jenkins.io/doc/book/pipeline/shared-libraries/ is unfortunate. By definition a “library” is “shared”; why else would you create a library? The [plugin itself](https://github.com/jenkinsci/pipeline-groovy-lib-plugin/) uses the term _Pipeline Groovy library_.

Reply:
Updated to mention both terminologies, hopefully that makes it clearer to both populations of readers.

OTOH, I often see (and use) the acronym JSL though, but never saw a PGL...

Reply:
Although I would say that code-sharing is one aspect. Code offloading (into a separate compilable unit) is another :)
Some organizations might want the looks of a declarative pipeline for some job with their custom logic behind it, so the logic is in a library (trusted, un-sandboxed and all) while the pipeline seems to call a few steps with a clean interface. But it may be the only consumer (the not-quite-sharing aspect of this).